tests/Integration/Classes/CartRuleTest.php

```diff
@@ -0,0 +1,259 @@
+
+        $customerCartRules = CartRule::getAllCustomerCartRules(
+            (int) $this->dummyCustomer->id
+        );
+
+        $this->assertEquals(2, count($customerCartRules));
+    }
+
+    /**
+     * Test sorting of the CartRules
+     *
+     * Target sort order:
+     * - Active CartRules for customer
+     * - Active CartRules for everyone
+     * - Inactive CartRules for customer
+     * - Inactive CartRules for everyone
+     */
+    public function testSortingOfTheAvailableCustomerCartRules(): void
+    {
+        // inactive customer's rule
+        $inactiveCustomerRule = $this->createDummyCartRule(false, (int) $this->dummyCustomer->id, false);
+
+        // inactive global rule
+        $inactiveGlobalRule = $this->createDummyCartRule(false, 0, false);
+
+        // active global rule
+        $activeGlobalRule = $this->createDummyCartRule(true, 0, false);
+
+        // active customer's rule
+        $activeCustomerRule = $this->createDummyCartRule(true, (int) $this->dummyCustomer->id);
+
+        $customerCartRules = CartRule::getAllCustomerCartRules(
+            (int) $this->dummyCustomer->id
+        );
+
+        $this->assertEquals(
+            [
+                $activeCustomerRule->id,
+                $activeGlobalRule->id,
+                $inactiveCustomerRule->id,
+                $inactiveGlobalRule->id,
+            ],
+            array_column($customerCartRules, 'id_cart_rule')
+        );
+    }
+
+    /**
+     * @param bool $active
+     * @param int $customerId
+     * @param bool $code
+     *
+     * @return CartRule
+     */
+    public function createDummyCartRule(
+        bool $active,
+        int $customerId,
+        bool $code = true
+    ): CartRule {
+        $randomNumber = rand(999, 9999);
+        $cart_rule = new CartRule();
+        $cart_rule->code = $code ? 'TEST_CART_RULE_' . $randomNumber : '';
+        $cart_rule->name = [
+            $this->defaultLanguageId => 'Test Cart Rule #' . $randomNumber,
+        ];
+        $cart_rule->id_customer = $customerId;
+        $cart_rule->free_shipping = true;
+        $cart_rule->quantity = 1;
+        $cart_rule->quantity_per_user = 1;
+        $cart_rule->minimum_amount_currency = $this->configuration->get('PS_CURRENCY_DEFAULT', null, ShopConstraint::allShops());
+        $cart_rule->reduction_currency = $this->configuration->get('PS_CURRENCY_DEFAULT', null, ShopConstraint::allShops());
+        $cart_rule->date_from = date('Y-m-d H:i:s', time());
+        $cart_rule->date_to = date('Y-m-d H:i:s', time() + 24 * 36000);
+        $cart_rule->active = $active;
+        $cart_rule->add();
+
+        return $cart_rule;
+    }
+
+    /**
+     * @return Customer
+     */
+    public function createDummyCustomer(): Customer
+    {
+        $customer = new Customer();
+        $customer->firstname = 'Jenna';
+        $customer->lastname = 'Doe';
+        $customer->email = 'pub+jenna+' . rand(0, 99) . '@prestashop.com';
```

Comment:
I know there is very little chance that it happens, but this doesn't garanti the uniqueness of the email 😅

Reply:
Woot, I missed this one :scream: